src/web/utils/AppVisibilityUtils.ts

```diff
@@ -0,0 +1,96 @@
+/**
+* AppVisibilityUtils.ts
+*
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT license.
+*
+* Web-specific helpers for firing focus/activity related events
+*/
+
+import SubscribableEvent from 'subscribableevent';
+
+const idleTimeInMs = 10 * 1000;
+
+export class AppVisibilityUtils {
+    private _isIdle = false;
+    private _timers: Array<number> = [];
+    
+    onFocusedEvent = new SubscribableEvent<() => void>();
+    onBlurredEvent = new SubscribableEvent<() => void>();
+    onAppForegroundedEvent = new SubscribableEvent<() => void>();
+    onAppBackgroundedEvent = new SubscribableEvent<() => void>();
+    onIdleEvent = new SubscribableEvent<() => void>();
+    onWakeUpEvent = new SubscribableEvent<() => void>();
+
+    constructor() {
+        window.addEventListener('focus', this._onFocus);
+        window.addEventListener('blur', this._onBlur);
+        document.addEventListener('visibilitychange', this._onAppVisibilityChanged);
+
+        this._trackIdleStatus();
+    }
+
+    hasFocusAndActive() {
+        return document.hasFocus() && !this._isIdle;
+    }
+
+    hasFocus() {
+        return document.hasFocus();
+    }
+
+    private _trackIdleStatus () {
+        document.addEventListener('mousemove', this._wakeUpAndSetTimerForIdle);
+        document.addEventListener('keyup', this._wakeUpAndSetTimerForIdle);
+        document.addEventListener('touchstart', this._wakeUpAndSetTimerForIdle);
+        document.addEventListener('scroll', this._wakeUpAndSetTimerForIdle);
+
+        this._wakeUpAndSetTimerForIdle();
+    }
+
+    private _wakeUpAndSetTimerForIdle = () => {
+        this._timers.map(clearTimeout);
+
+        if (!this.hasFocus()) {
+            return;
+        }
+
+        if (this.hasFocus() && this._isIdle) {
+            this._onWakeUp();
+        }
+
+        this._timers.push(setTimeout(() => {
+            if (this.hasFocus()) {
+                this._onIdle();
+            }
+        }, idleTimeInMs));
+    }
+
+    private _onFocus = () => {
+        this.onFocusedEvent.fire();
+        this._wakeUpAndSetTimerForIdle();
+    }
+
+    private _onBlur = () => {
+        this.onBlurredEvent.fire();
+        this._onIdle();
```

Comment:
Same here. The event should be fired only after _onIdle is called.

Reply:
By calling _onIdle, you're going to fire the "onIdle" event as well as the "onBlurred" event. Is that intentional? Wouldn't it be better to fire only the "onBlurred" event here since we're not really hitting an idle condition?

Reply:
I was debating about this. I copied the way ifvisible works. 

Essentially my thinking was, there could be listeners for only wakeup and idle events. And the app is idle when not blurred. Ifivisble does the same as well - when it goes to blur, it also fires idle event. Which i think makes sense - in which case each event has correct information about the status. 

Same goes for wakeup - current i fire "wakeup" when app is focused and when there is activity after idle. 

And same for app foreground and app background. We fire blur event, when app is backgrounded as well. And focused even when app is foregrounded. So they are independent pair of events. 

What do you think?

Reply:
OK, sounds reasonable. Just wanted to make sure it was intentional.